_drafts/debug-vs-release-builds.md

```diff
@@ -0,0 +1,202 @@
+
+# compare the two files; they're bit-for-bit identical
+❯ diff hello-no-debug hello-with-debug && echo 'identical!'
+identical!
+```
+
+**Summary:**
+
+- debug info should always be enabled!
+
+## Compiler optimization level
+
+Setting a lower (or zero) compiler optimization level can help in some
+step-debugging scenarios. It prevents the compiler from optimizing away
+variables when they're no longer relevant, or rearranging/combining fragments of
+code during optimization passes.
+
+Optimization can make step-debugging confusing, but the meaningful
+information, such as data structures, variables, and the call stack, is present.
+I've never hit a roadblock due to compiler optimization in step-debugging. In the worst
+case, I need to peek at a few assembly instructions to see what's going on, but
+that's rare.
+
+There's a HUGE downside to having different optimization levels for debug vs.
+release builds: we end up with a **different** executable:
+
+- sizes will likely be different: stack or heap usage (affecting program
+  stability!), code space (affecting flash memory allocation, OTA)
+- different machine code executed by the CPU: this can impact **timing** (often
+  where the nastiest bugs show up) and actual **program validity**, e.g.
+  assumptions you are making about certain spots in your program may no longer
+  be valid with different optimization levels (the classic one is instruction
+  reordering causing concurrent access bugs!)
+
+Because your debug (testing) build is not running the same code, you may see
+these bugs only show up during field test, rather than when the device is on
+your desk and connected to a debugger!
+
+It's also generally not desirable to adjust compiler optimization level when a
+project is nearing time to ship, to decrease risk. If the debug + release builds
+are using the same optimization settings, it's less likely to hit a surprise
+out-of-space issue that can stop a project in its tracks 😱.
+
+**Summary:**
+
+- use the same optimization settings for all builds to decrease risk
+
+## `NDEBUG` preprocessor definition
+
+Often these Debug/Release builds will have some preprocessor defined symbols
+such as:
+
+- `DEBUG=1` (debug enabled)
+- `NDEBUG=0` (indicating this is NOT a non-debug build)
+
+Usually these flags will be used to compile in features like this to the Debug
+builds:
+
+- enable an external CLI for testing (this is so useful in production devices! just
+  secure it properly)
+- disable runtime asserts (assuming your system can tolerate a reset due to an
+  `ASSERT` condition, these are stupendously valuable in production builds- see
+  [this post]({% post_url 2019-11-20-cortex-m-hardfault-debug %}) for more
+  details)
+- changing default log levels, i.e. enabling `INFO` or `DEBUG` levels- this can
+  impact timing or memory/bandwidth usage, so generally it's preferable to stick
+  to one level and only increase in select cases when you need extra information
```

Comment:
So the recommendation is to use a level of logging  in development that is acceptable in production? You might want to clarify this.

I am not sure that is always true. I think there is a lot of value in collecting extra logs in development and field testing, although you cannot afford to keep the extra logs in production because of flash wear. What would you say to this argument?

Reply:
very good point, I'll include it! I like the idea of runtime configuration here but not always practical, so this is a great example of "yes, production builds may want to be different"